### PR TITLE
accept YYYY-MM-DD date format in stdext

### PR DESCRIPTION
**Pull request by lippirk**

### CODE REVIEW

File: lib/xapi-stdext-date/date.ml [hunk 1]
```diff
@@ -12,18 +12,29 @@
...
+(** this dash type is used to store information about whether
+    the input string contained dashes or not (e.g. it is false when
+    `of_float` is called)
+  * we remember the dashes only where possible:
+     - `"yyyy-mm-ddTxx" |> of_string |> to_string = "yyyy-mm-ddTxx"
+  *  - `"yyyy-mm-ddTxx" |> of_string |> to_float |> of_float |> to_string = yyyymmddTxx`
+  *)
+type dash = bool
+
+type iso8601 = (string * dash)
```
Comment (edwintorok) [inline lib/xapi-stdext-date/date.ml:24]: Since this time is abstract, can't we replace it with a Ptime, so that we know we have parsed it correctly?
The code below does a lot of juggling with offsets between UTC and localtime, where I think ptime could parse it independently of timezone. For backwards compatibility we could convert something without dashes into something that Ptime expects.

If we want to retain backwards compatibility wrt to where we raise exceptions you could have:
```
type iso8601 = {
   ptime:(Ptime.t * Ptime.t tz_offset_s option, `Msg of string) Result.t;
   orig: string;
   compat: bool
}
```

This would allow you to carry the date unchanged across of_string/to_string without raising exceptions (like the old code did), and raise exceptions only at the time of conversion to float, or at the assert utc.

File: lib/xapi-stdext-date/date.ml [hunk 2]
```diff
@@ -12,18 +12,29 @@
...
 
-type iso8601 = string
+(** this dash type is used to store information about whether
+    the input string contained dashes or not (e.g. it is false when
+    `of_float` is called)
+  * we remember the dashes only where possible:
+     - `"yyyy-mm-ddTxx" |> of_string |> to_string = "yyyy-mm-ddTxx"
+  *  - `"yyyy-mm-ddTxx" |> of_string |> to_float |> of_float |> to_string = yyyymmddTxx`
+  *)
+type dash = bool
```
Comment (psafont) [inline lib/xapi-stdext-date/date.ml:22]: Why is this needed, wouldn't changing the of_string function be enough to accept the format with dashes?

Comment (lippirk) [reply lib/xapi-stdext-date/date.ml:22]: it's to ensure that `of_string . to_string` is the identity function

i imagine that it isn't necessary, but if somebody saved one format to the database, for example, they might expect the same format to come out again

File: lib/xapi-stdext-date/date.ml (hunk 1 above)
Comment (lippirk) [reply lib/xapi-stdext-date/date.ml:24]: that sounds like a better idea :) i will implement that

File: lib/xapi-stdext-date/date.ml (hunk 2 above)
Comment (psafont) [reply lib/xapi-stdext-date/date.ml:22]: Can that property can be kept on a serialization + deserialization? If not I wouldn't worry about it, that's the main use-case of the library.

File: lib/xapi-stdext-date/date.ml [hunk 3]
```diff
@@ -12,19 +12,9 @@
  * GNU Lesser General Public License for more details.
  *)
 
-type iso8601 = string
+(* ==== RFC822 ==== *)
```
Comment (psafont) [inline lib/xapi-stdext-date/date.ml:15]: You might want to use `(**` for these comments to make them appear in the docs

Comment (lippirk) [reply lib/xapi-stdext-date/date.ml:15]: this was more meant to help separate functions related to this rfc822 type, and the iso8601 type whilst looking at the file

File: lib/xapi-stdext-date/date.ml [hunk 4]
```diff
@@ -38,35 +28,86 @@ let rfc822_of_float x =
...
+(* ==== ISO8601/RFC3339 ==== *)
+type old_time = string
+type error = string
+type p_time = Ptime.t * Ptime.tz_offset_s option
+
+(** the name doesn't make much sense anymore, but is kept for compatibility reasons
+  * if string is received in rfc3339 format, timezone will be preserved between to/from string
+    but not once converted to a float
+*)
+type iso8601 =
```
Comment (lindig) [inline lib/xapi-stdext-date/date.ml:40]: I like the design but like these names better:
```
type iso8601 =
  | Time of p_time
  | Legacy of string * error
```

I am a bit confused why an error is part of this data type. Should this not be orthogonal?

Comment (psafont) [reply lib/xapi-stdext-date/date.ml:40]: I think this is due to a mismatch between parsing and validation that happens in the previous design.

There was no parsing whatsoever, allowing any string to be converted, and then there was a validation further down the line using `assert_utc`.

I'd rather parse the string into a utc-based timedate (Ptime, float) and fix what this change of behaviour breaks rather than trying to maintain all the information to have `of_string . to_string = id`

File: lib/xapi-stdext-date/date.ml [hunk 5]
```diff
@@ -38,35 +28,86 @@ let rfc822_of_float x =
...
+  | Old of old_time * error (* iso8601 - accepts both YYYYMMDD & YYYY-MM-DD *)
+
+ let of_string x =
+  let r = Ptime.of_rfc3339 x in
+  match r with
+  | Ok (t, tz, _) -> P_time (t, tz) (** prefer to use ptime
+                                      * it will only accept dashes *)
+  | Error e -> match Ptime.rfc3339_error_to_msg r with
+               | Error (`Msg err_msg) -> Old (x, err_msg)
+               | Ok x -> failwith "impossible" (* ptime providing bad api *)
```
Comment (lindig) [inline lib/xapi-stdext-date/date.ml:51]: I'd suggest to use `"impossible: "^__LOC__` - so we don't depend on the stack trace.

Comment (lippirk) [reply lib/xapi-stdext-date/date.ml:51]: i realised that you can simply the code to avoid this

**Comment by lippirk** [comment]
i'll run a BST to see if that catches any bugs

**Comment by lippirk** [comment]
BST giving errors because XRT makes assumptions about how it parses xapi datetimes - a PR is open for that

File: lib/xapi-stdext-date/date.mli [hunk 6]
```diff
@@ -24,7 +24,8 @@ val of_float : float -> iso8601
 (** Convert date/time to a float value: the number of seconds since 00:00:00 UTC, 1 Jan 1970. *)
 val to_float : iso8601 -> float
 
-(** Convert date/time to an ISO 8601 formatted string. *)
+(** Convert date/time to an ISO 8601 formatted string.
+  * Does not accept a timezone annotated datetime *)
```
Comment (psafont) [inline lib/xapi-stdext-date/date.mli:28]: Not sure why this line got added, did it make sense before switching to the latest approach?

Comment (lippirk) [reply lib/xapi-stdext-date/date.mli:28]: i meant to add this to `of_string`
